### ISSUE TITLE
Re-purpose / reuse root for SSR hydration / prerendering.

### ISSUE THREAD

**Issue by jbucaran** [hyperapp/hyperapp#319]
Our [current hydration implementation](https://github.com/hyperapp/hyperapp/blob/69676ce2c8d4b0c1fc167b37386226287ba5db31/src/app.js#L47) is broken! We are simply avoiding the duplication of your SSR nodes, but we are still patching everything from scratch. 😆

We could improve it this way:

```js
function hydrate(element, map) {
  return element
    ? {
        tag: element.tagName.toLowerCase(),
        data: {},
        children: map.call(element.childNodes, function(element) {
          return element.nodeType === 3
            ? element.nodeValue
            : hydrate(element, map)
        })
      }
    : element
}
```

Unfortunately, this still does not take into account empty text nodes, which if not trimmed, would yield a tree that is not consistent with the **new** tree returned by the view, causing patch to do more work than it needs if we were hydrating correctly.

A more complex solution could be explored, but then how much is enough and can this be moved outside core?

I've been thinking about this today and considered many alternatives, and here is my proposal.

---

Re-purpose `root` for SSR-use for what-would-be the old element during the initial render.

I originally chose the name `root` because it was short and simple, but now I think that `root` should describe your application's top element: `<main>`, `<div>`, etc., and NOT the HTML element to act as a container / host of your application tree (which is what we do now).

### Apps not using `root`.

Everything will continue to work as usual. 🎉 

### Apps using `root`.

Say you have this HTML:

```html
<body>
  <div id="app"></div>
</body>
```

and used `root` like this:

```js
app({
  view: () => <h1>Hello.</h1>,
  root: document.getElementById("main")
})
```

Now your app will render **overwriting / obliterating** `<div id="app">`, so the final HTML will look like this:

```html
<body>
  <h1>Hello.</h1>
</body>
```

This is similar to how React behaves.

### SSR

Using `root` for SSR pages would allow us to remove this part from core.

```js
element = document.querySelector("[data-ssr]")
```

And use instead:

```js
app({
  ...,
  root: document.querySelector("[data-ssr]")
})
```

Or a different data attribute; it's up to you.

Finally, the old node for the initial render. Instead of introducing a new property, I propose using the `load` event, currently unused.

```jsx
app({
  events: {
    load(state, actions, root) {
      return myOwnHydrationFormula(root)
    }
  }
})
```

so we can assign that result value to the node inside core like this:

```js
node = emit("load", root)
``` 

Using a mixin for that would work as well. Obviously, we'd call that mixin `Hydrator`!

```jsx
app({
  mixins: [Hydrator]
})
```

### SSR Example

A full SSR page example would work like this:

```html
<html>

<head>
  <script defer src="bundle.js"></script>
</head>

<body>
  <main data-hyperapp-ssr>
    <h1>My app.</h1>
  </main>
</body>

</html>
```

```jsx
app({
  view: () => 
     <main>
        <h1>My app.</h1>
     </main>,
  mixins: [Hydrator],
  root: document.querySelector("[data-hyperapp-ssr]"),
})
```

/cc @andyrj @ngryman @SkaterDad @lukejacksonn @zaceno

**Comment by jbucaran** [comment]
By the way, with this now it's also possible to do what until now was impossible. Use the `<body>` inside the view.

<pre>
view: () =>
  <b>&ltbody&gt</b>
     ...
  <b>&lt/body&gt</b>  
</pre>

**Comment by zaceno** [comment]
I really liked @Swizz 's idea of providing emit to the load event, so by default, the app would return emit. 

I cant say for sure which is most rational/useful to pass emit("load", ...) - emit or root? But my hunch and personal opinion is that it should be emit. Esp if it's possible to solve SSR anyway, with a mixin.

**Comment by jbucaran** [comment]
@zaceno The `load` event must receive the root element so that you can hydrate if you want. An alternative event could be devised as well.

By the way, this has nothing to do with the return value of the app() call. We can continue to return `emit` as usual.

**Comment by andyrj** [comment]
I still think you would need node, not root passed to the load event @jbucaran .

Worrying about the extra work of hydration seems extreme since it only ever runs once.  But I am completely onboard with moving this out of core if that's what you want.

**Comment by zaceno** [comment]
@jbucaran Isn't the idea that app will return the result of the load-event? Or am I getting old and new app event names mixed up?

**Comment by jbucaran** [comment]
@zaceno I should have made it clearer that we can't reuse `load` for that anymore (unless we introduce an event dedicated only for hydration).

So, hyperapp 0.11.0 will continue to return `emit`.

**Comment by zaceno** [comment]
Oh, yes, I misunderstood - thanks for clarifying!

**Comment by jbucaran** [comment]
> I still think you would need node, not root passed to the load event @jbucaran.

I don't see why that would be needed, but can you provide an example?

> Worrying about the extra work of hydration seems extreme since it only ever runs once. But I am completely onboard with moving this out of core if that's what you want.

Is it really extreme?

### Before

```jsx
app({
  view: () => 
     <main>
        <h1>My app.</h1>
     </main>
})
```

### After

```jsx
app({
  view: () => 
     <main>
        <h1>My app.</h1>
     </main>,
  mixins: [Hydrator],
  root: document.querySelector("[data-hyperapp-ssr]"),
})
```

The new approach has more code and requires more user intervention, but it also makes it very obvious and at the same time offers more flexibility as it leaves it up to you to implement hydration.

I like how it forces the user to target the root `[data-hyperapp-ssr]` **while at the same time** making this property more consistent with what other frameworks (React) do and without breaking existing apps not using the prop (only requiring little changes to those using it).

**Comment by SkaterDad** [comment]
:+1: 

I'm in favor of this idea.  Replacing the root node makes more sense, and it's one less useless `<div>` wrapper around an app.  Also good that we can now specify what attributes/id to target.

However you decide to move the hydration out of core, that's probably for the best also.  No reason to ship code in core that 75% of your users won't be needing (number pull out of thin air).

**Comment by jbucaran** [comment]
@SkaterDad It also lets you render a `<body>` tag. 😄

**Comment by jbucaran** [comment]
By the way, I want to clarify that I **do** plan to use SSR in real apps. But I don't see why we should include it in core when we could use a mixin just as well.

**Comment by andyrj** [comment]
@jbucaran the node variable is required for hydration in general, but root is already declared by the user so can be sent to the mixin by the user rather trivially.  Without node the result of your Hydrator.events.load() can't provide the app with the vnode skeleton.  Maybe it is accessible via some js voodoo that I am missing but app is basically a closure and the node var is hidden from the mixin unless I am missing something.

**Comment by jbucaran** [comment]
@andyrj Actually events can return values. I wrote this above, but here it is again for convenience:

```jsx
app({
  events: {
    load(state, actions, root) {
      return myOwnHydrationFormula(root)
    }
  }
})
```

**Comment by jbucaran** [comment]
\cc **everyone** Now, when I refer to `root`, this is no longer the previous odd meaning of application tree container, but what a root _is_, your application top-level element.

**Comment by andyrj** [comment]
I apologize for missing that, I see what you are going to do, you are right node is not necessary if you do that.  Still getting my coffee this morning heh.

**Comment by jbucaran** [comment]
@andyrj With this new information, how do you feel about the proposal? You are who originally brought SSR to HyperApp, so I value your opinion here! 🙏

**Comment by andyrj** [comment]
But I still think adding root to the load event is un-necessary.
```js
var root = document.querySelector("[data-hyperapp-ssr]")
app({
  view: ...,
  mixin: Hydrator(root),
  root
})
```

**Comment by jbucaran** [comment]
@andyrj Yeah, sure, it's just for convenience.

**Comment by andyrj** [comment]
I do have one more concern with this proposal.  If we overwrite the target node during init that will cause issues for my current webpack hmr setup.  If I target a node with id="root" and our app obliterates it, it does make the final output "cleaner", but then in module.hot I have no element to target for the re-initialization of app().   :disappointed:

**Comment by jbucaran** [comment]
> overwrite the target node during init that will cause issues for my current webpack hmr setup

I am sure there must be a workaround, since overwriting the target node is how, e.g., React behaves.

**Comment by jbucaran** [comment]
@andyrj > overwrite the target node during init that will cause issues for my current webpack hmr setup

I am sure there must be a workaround, since overwriting the target node is how, e.g., React behaves.

**Comment by jbucaran** [comment]
> @andyrj then in module.hot I have no element to target for the re-initialization of app()

Can you share some code with the problem? I have no idea what you mean.

**Comment by andyrj** [comment]
I haven't messed with react official for some time, but last I checked it left the render target in tact, as do inferno and preact.

I will use your proposed example above as a base.

```html
<body>
  <div id="app"></div>
</body>
```

```js
/* index.js */
let target = document.getElementById('app');
app({
  view: () => 
     <main>
        <h1>My app.</h1>
     </main>,
  mixins: [Hydrator],
  root: target
})

if (module.hot) {
  module.hot.accept("./app", () => {
    // blows up below because target no longer exists in the document if we obliterate it...
    app(window.state, target)
  })
}
```

**Comment by jbucaran** [comment]
@andyrj Are you sure you don't know how to fix this? 😏

**Comment by andyrj** [comment]
I don't see a good work around, I would need to document.body.innerHTML="<div id="app"></div>" before the call to app() in module.hot, so back to the flickering <seizure> lol.  Don't change what you want to do based around my issue, I am just saying it will cause issues for my use case of wanting ssr with hmr without flickering.  I can always keep my private fork going...  I still despise having to add data attributes and not being able to SSR to body tag to keep the strange append use case so I will likely always be using my private fork.  But I could see this being strange to others that also use webpack-hmr.

**Comment by jbucaran** [comment]
@andyrj I can see two clean ways and a dirty way. The dirty way involve calling getElementById("app") twice, but let's avoid that.

### Using rehydration!

```html
<body>
  <div id="app">
   <!-- rest of the tree, or only the top-level element!
    -->
  </div>
</body>
```
```jsx
export default (state, target) => {
  app({
    view: () => 
       <div id="app">
          <h1>My app.</h1>
       </div>,
      root: target
  })
};
```

If your Hydrator is working correctly, then HyperApp's patch() should not end up replacing the element, but doing exactly what hydration is meant for and reuse the root and children. It will only call updateElementData and return since there is no data to update. Then it would repeat the same process for every child if your SSR page contains child nodes.

### Simulating the previous behavior (no-hydration)

```html
<body>
  <main> <!-- will not be touched -->
      <div id="app"></div> <!-- will be replaced -->
  </main>
</body>
```

```jsx
const root = document.getElementById("app")
```

**Comment by MatejMazur** [comment]
@andyrj @jbucaran Just a quick notice: I'm trying HMR and it's working without flickering at least if you don't use SSR (I haven't used SSR yet so I don't know). But now I'm testing HMR usage in one game project with the latest changes in API. So webpack-hmr is perfectly doable for client side dev. 👍

**Comment by andyrj** [comment]
LGTM!  Looks great on my projects this will let me get rid of my fork I have been using.  :100: :+1:

**Comment by jbucaran** [comment]
Implemented in #317.
